### PR TITLE
ENH: Carlson symmetric elliptic integrals.

### PR DESCRIPTION
**Pull request by duetosymmetry**
#### Reference issue

Supersedes #10504, which itself supersedes #10304

#### What does this implement/fix?

Original text from @congma:

> This changeset implements the Carlson symmetric elliptic integrals
R{C,D,F,G,J} as ufuncs in scipy.special. The functions
ellipr{c,d,f,g,j} takes real or complex floating-point arguments and
perform the calculation based on the iterated duplication algorithm and
series expansion.

> The underlying extension is written as C++ function templates. This
supersedes PR #10304 where the same functionalities were
implemented in C. The latter relied on C11 features currently not
well-supported on some platforms.

#### Additional information

Please see the extended discussion which took place at #10504. The original author, @congma (to whom all credit is due), has not responded since Nov. 2020. Several reviewers deemed this PR ready. I have rebased it here (with minor additional documentation, and some additional tests) since master has diverged so far since then.

Original reviewers: @rgommers @arokem @duetosymmetry @sturlamolden @mckib2 @pv.

I recommend that a request for additional testing for `elliprj` be opened as an issue, which was the last issue that @congma commented about.

I also recommend that @mckib2 make their desired C++ improvements as a separate PR.

There are too many commits here, I will squash when so desired. I just wanted to preserve the original commit history when first opening this PR.

### CODE REVIEW

**Comment by duetosymmetry** [comment]
Fixups for failing lint checks (which is more stringent than when the earlier PR was opened)

**Comment by duetosymmetry** [comment]
Rebase onto updated master. Upstream made a related change to handling of scipy/special/sf_error.c for C++ extensions, which the original author here also made, leading to a merge conflict. The upstream approach was better. I have rewritten the history here to simplify the back-and-forth changes that happened in scipy/special/setup.py .

File: scipy/special/_add_newdocs.py [hunk 1]
```diff
@@ -2210,6 +2251,341 @@ def add_newdoc(name, doc):
...
+
+    Notes
+    -----
+    RC is a degenerate case of the symmetric integral RF: ``elliprc(x, y) ==
+    elliprf(x, y, y)``. It is an elementary function rather than an elliptic
+    integral.
+
+    The code implements Carlson's algorithm based on the duplication theorems
+    and series expansion up to the 7th order. [2]_
+
```
Comment (rgommers) [inline scipy/special/_add_newdocs.py:2295]: Could you add
```

.. versionadded:: 1.8.0
```
here and in the other docstrings?

**Review by rgommers — COMMENTED** [review]
That plan looks good, thanks @duetosymmetry. Squashing commits seems reasonable; you can keep multiple if that makes sense - it's a large PR so even the current state with 35 commits would be acceptable, but squashing commits with repeated changes to the same code and small cleanup commits would be nice.

I had another look at the docs, they look polished. Started CI, the jobs that are finished already all seem happy.

File: scipy/special/_add_newdocs.py (hunk 1 above)
Comment (duetosymmetry) [reply scipy/special/_add_newdocs.py:2295]: Sure thing! Addressed in ed17333.

**Comment by duetosymmetry** [comment]
Failures are only due to timeouts after 60m (during publishing of code coverage).

**Comment by duetosymmetry** [comment]
The commit history naturally squashed down to three commits:
1. When @congma originally declared the code feature-complete
2. When they addressed reviewer comments (and their own testing improvements)
3. My additional docs/test at the end.

Should be ready to merge.

**Comment by duetosymmetry** [comment]
The one test marked as failing ran for 60m and timed out (99% of tests had completed). Not sure why this particular configuration is slow enough to hit the time limit when all the other ones finish in under an hour.

File: scipy/special/tests/test_basic.py [hunk 2]
```diff
@@ -1581,6 +1582,187 @@ def test_ellipeinc_2(self):
...
+                                  1.6490011662711,
+                                  0.94148358841220,
+                                  1.8260115229009+1.2290661908643j,
+                                  -0.61127970812028-1.0684038390007j,
+                                  0.24723819703052,    # Cauchy principal value
+                                  -0.12711230042964])  # Caucny principal value
+        for i, arr in enumerate(args):
+            assert_allclose(elliprj(*arr), expected_results[i])
+
+    @pytest.mark.xfail(reason="Insufficient accuracy on 32-bit")
```
Comment (tylerjereddy) [inline scipy/special/tests/test_basic.py:1707]: `xfail` accepts a condition as the first argument, so you could still run this in 64-bit cases

Reviewers seem positive on getting this in though so I won't block it on that, which can be fixed up later I guess.

I also noted in a scan through of the code that no old tests have really been modified, only new tests added, which increases safety a bit. The CI also reports decent test coverage. I'm admittedly leaning on the described thorough review history of the work.

**Comment by tylerjereddy** [comment]
thanks

**Comment by rgommers** [comment]
Awesome to see this merged finally, thanks @congma, @duetosymmetry, @arokem and everyone else who contributed!

@duetosymmetry would you mind adding a short description of the new feature to https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.8.0?

**Comment by duetosymmetry** [comment]
Release notes updated. Thanks, maintainers!

**Comment by WarrenWeckesser** [comment]
Nice addition to SciPy!

Now engaging `nag` mode...

Don't forget that including examples in new functions is bullet item 4 of the [Checklist before submitting a PR](https://docs.scipy.org/doc/scipy/reference/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr).  
Could someone familiar with these new functions add some illustrative examples to the docstrings of `elliprc`, `elliprd`, `elliprf`, and `elliprj`?  We want the number of functions found by the script in https://github.com/scipy/scipy/issues/7168 to go down, not up!

**Comment by duetosymmetry** [comment]
@WarrenWeckesser sorry, I am new to contributing to scipy, and did not read the checklist before taking over the PR. The only examples I know are quite niche, but I saw some nice general examples [in the mpmath docs](https://mpmath.org/doc/current/functions/elliptic.html#ellippi). If you create an issue and assign it to me, I will start to work on it after I finish grading finals ;)

**Comment by WarrenWeckesser** [comment]
@duetosymmetry, there is no need to apologize! That checklist can be hard to find, even if you know it exists.  It is primarily the responsibility of the reviewers to keep on eye on things like this.

I don't think we need an issue for this, just create a pull request when you have chance.  Thanks!

**Pull request merged**